src/services/jsonSchemaService.ts

```diff
@@ -433,20 +433,54 @@ export class JSONSchemaService implements IJSONSchemaService {
 			return current;
 		};
 
-		const merge = (target: JSONSchema, sourceRoot: JSONSchema, sourceURI: string, refSegment: string | undefined): void => {
+		const merge = (target: JSONSchema, section: any): void => {
+			for (const key in section) {
+				if (section.hasOwnProperty(key) && !target.hasOwnProperty(key)) {
+					(<any>target)[key] = section[key];
+				}
+			}
+		};
+
+		const mergeByJsonPointer = (target: JSONSchema, sourceRoot: JSONSchema, sourceURI: string, refSegment: string | undefined): void => {
 			const path = refSegment ? decodeURIComponent(refSegment) : undefined;
 			const section = findSection(sourceRoot, path);
 			if (section) {
-				for (const key in section) {
-					if (section.hasOwnProperty(key) && !target.hasOwnProperty(key)) {
-						(<any>target)[key] = section[key];
-					}
-				}
+				merge(target, section);
 			} else {
 				resolveErrors.push(localize('json.schema.invalidref', '$ref \'{0}\' in \'{1}\' can not be resolved.', path, sourceURI));
 			}
 		};
 
+		const isSubSchemaRef = (refSegment?: string): boolean => {
+			// Check if the first character is not '/' to determine whether it's a sub schema reference or a JSON Pointer
+			return !!refSegment && refSegment.charAt(0) !== '/';
+		};
+
+		const reconstructRefURI = (uri: string, fragment?: string, separator: string = '#'): string => {
+			return normalizeId(`${uri}${separator}${fragment}`);
+		};
+
+		// To find which $refs point to which $ids we'll keep two maps:
+		// One for '$id' we expect to encounter (if they exist)
+		// and one for '$id' we have encountered
+		const pendingSubSchemas: Record<string, JSONSchema[]> = {};
+		const resolvedSubSchemas: Record<string, JSONSchema> = {};
+
+		const tryMergeSubSchema = (uri: string, target: JSONSchema) : boolean => {
+			if (resolvedSubSchemas[uri]) { // subschema is resolved, merge it to the target
+				merge(target, resolvedSubSchemas[uri]);
+				return true; // return success
+			} else { // This subschema has not been resolved yet
+				if (!pendingSubSchemas[uri]) {
+					pendingSubSchemas[uri] = [];
+				}
+
+				// Remember the target to merge later once resolved
+				pendingSubSchemas[uri].push(target);
+				return false; // return failure - merge didn't occur
+			}
+		};
+
```

Comment:
This is the heart of the solution. It basically keeps two maps for subschema references.  One for those subschemas that we may or may not encounter later and the latter for those subschemas we have resolved. 

The `tryMergeSubSchema` merges already resolved subschemas into the targets that `$ref` them, or keeps a record of the target object against the full URI of the subschema for later use.